### PR TITLE
cuda 8 api uses cudaDataType for 16bit float type

### PR DESCRIPTION
**Pull request by kashif**

### CODE REVIEW

File: tensorflow/stream_executor/cuda/cuda_blas.cc [hunk 1]
```diff
@@ -1680,10 +1680,25 @@ bool CUDABlas::DoBlasGemm(
   return DoBlasInternal(
       dynload::cublasSgemmEx, stream, true /* = pointer_mode_host */,
       CUDABlasTranspose(transa), CUDABlasTranspose(transb), m, n, k, &alpha,
-      CUDAMemory(a), CUBLAS_DATA_HALF, lda,
-      CUDAMemory(b), CUBLAS_DATA_HALF, ldb,
-      &beta,
-      CUDAMemoryMutable(c), CUBLAS_DATA_HALF, ldc);
+      CUDAMemory(a),
```
Comment (vrv) [inline tensorflow/stream_executor/cuda/cuda_blas.cc]: I wonder if it makes sense defining our own "half type" somewhere in another header that we typedef once, so we centralize the definition of the symbol.  Thoughts @zheng-xq  ?

Comment (kashif) [inline tensorflow/stream_executor/cuda/cuda_blas.cc]: yes or I can also `#define CUBLAS_DATA_HALF CUDA_R_16F` for `CUDA_VERSION >= 8000` which might make the code look a bit cleaner...

Comment (vrv) [inline tensorflow/stream_executor/cuda/cuda_blas.cc]: That seems fine to me

File: tensorflow/stream_executor/cuda/cuda_blas.cc [hunk 2]
```diff
@@ -25,6 +25,10 @@ limitations under the License.
 #define EIGEN_HAS_CUDA_FP16
 #endif
 
+#if CUDA_VERSION >= 8000
+#define CUBLAS_DATA_HALF CUDA_R_16F
+#endif
+
```
Comment (zheng-xq) [inline tensorflow/stream_executor/cuda/cuda_blas.cc]: This seems fine with me too. 

The downside is that it introduces an outdated symbol back into CUBLAS space, which might interact with other code when we move it around. I would second vrv@'s suggestion that it is better to have our own symbol that is conditionally defined.

I will leave it up to you whether you want to follow through in this CL.

**Comment by zheng-xq** [comment]
@tensorflow-jenkins test this please

**Comment by zheng-xq** [comment]
@vrv, @martinwicke, I've confirmed that with this patch, TensorFlow would build with Cuda 8. Feel free to merge this change. We can always improve it later.

**Comment by vrv** [comment]
test this please

File: tensorflow/stream_executor/cuda/cuda_blas.cc (hunk 2 above)
Comment (kashif) [inline tensorflow/stream_executor/cuda/cuda_blas.cc]: I will make the change in another pull request if that is ok?

**Comment by vrv** [comment]
@martinwicke @caisq our Jenkins is broken   

Merging since this change was verified by @zheng-xq

**Comment by kashif** [comment]
@zheng-xq  would something like:

```c++
#if CUDA_VERSION >= 8000
#define TF_CUDA_DATA_HALF CUDA_R_16F
#else
 #define TF_CUDA_DATA_HALF CUBLAS_DATA_HALF
#endif
```

and then I use `TF_CUDA_DATA_HALF` in the `cublasSgemmEx()`?

**Comment by zheng-xq** [comment]
@kashif, thanks for getting to this quickly. 

Please name it SE_CUDA_DATA_HALF, since this is stream-executor code, which TensorFlow only hosts a fork temporarily. Other than that, it looks fine.

**Comment by jendap** [comment]
@tensorflow-jenkins test this please

**Comment by jendap** [comment]
@kashif can you rename it please?

**Comment by jendap** [comment]
Thank you @kashif!

**Comment by kashif** [comment]
thank you! @jendap

**Pull request merged**